### PR TITLE
Add support for recaptcha's enterprise API

### LINKED ISSUE #390: Support for recaptcha enterprise?
Hi, 

Recaptcha enterprise uses `grecaptcha.enterprise.execute()` instead of `grecaptcha.execute()`. In this gem,  it's hard-coded to call `grecaptcha.execute()` https://github.com/ambethia/recaptcha/blob/8fd0ee1c6cd34810f55278b2aa2ddee0fd43483c/lib/recaptcha/helpers.rb#L189.

Is Enterprise support on the roadmap?

Thanks!
Danny

### PR DESCRIPTION
**Pull request by jcmfernandes**
Closes #390.

## Pre-Merge Checklist
- [x] CHANGELOG.md updated with short summary

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -506,6 +514,7 @@ Recaptcha.configuration.skip_verify_env.delete("test")
 Recaptcha.configure do |config|
   config.site_key  = '6Lc6BAAAAAAAAChqRbQZcn_yyyyyyyyyyyyyyyyy'
   config.secret_key = '6Lc6BAAAAAAAAKN3DRm6VA_xxxxxxxxxxxxxxxxx'
+  config.enterprise = true
```
Comment (grosser) [inline README.md:517]: default should be false though ?

**Review by grosser — APPROVED** [review]
looks nice and straight forward :D
lmk when you are happy with it

**Comment by jcmfernandes** [comment]
Thanks @grosser! 

This is WIP as the `verify_url` for the enterprise API is completely different.

I must confess that I was trying to open a PR in our own repo first and opened here by mistake :sweat_smile: I'll have this ready in the next few weeks, it's already working on my end :wink:.

File: lib/recaptcha/helpers.rb [hunk 2]
```diff
@@ -200,7 +201,7 @@ def self.to_message(_key, default)
           async function #{recaptcha_v3_async_execute_function_name(action)}() {
             return new Promise((resolve, reject) => {
               grecaptcha.ready(async function() {
```
Comment (pomartel) [inline lib/recaptcha/helpers.rb:203]:
```suggestion
             #{recaptcha_ready_method_name}(async function() {
```

File: lib/recaptcha/helpers.rb [hunk 3]
```diff
@@ -218,7 +219,7 @@ def self.to_message(_key, default)
         <script#{nonce_attr}>
           function #{recaptcha_v3_execute_function_name(action)}() {
             grecaptcha.ready(function() {
```
Comment (pomartel) [inline lib/recaptcha/helpers.rb:221]:
```suggestion
            #{recaptcha_ready_method_name}(function() {
```

**Comment by jcmfernandes** [comment]
Alright, I believe I'm done code-wise. We will be testing this for a few days here at @hopin-team and I'll then declare it up for review. We'll most likely be done with it this week.

**Comment by jcmfernandes** [comment]
And thank you @pomartel for the fixes! :raised_hands:

**Comment by pomartel** [comment]
Hey @jcmfernandes, thank you for your work on this PR. I just hit the limit of the free recaptcha plan so your PR came at the right time! I am using your branch and it seems to be working quite well. One thing I did notice though is that I get a lot more forms marked as possible bots. Did you also notice any change when switching to reCaptcha Enterprise? I wonder if it's something I didn't set correctly or if the Google algorithm might be different.

**Comment by jcmfernandes** [comment]
We have yet to migrate @pomartel. We also recently hit our limits and got the heads up from google about it. We plan to migrate this week hence why I haven't put this PR as ready for review. I'll keep you posted.

**Comment by jcmfernandes** [comment]
Alright, this is working nicely on our end. I'm declaring it ready for review :)

File: lib/recaptcha.rb [hunk 4]
```diff
@@ -121,4 +154,23 @@ def self.api_verification(verify_hash, timeout: nil)
...
+    query = URI.encode_www_form(query_params)
+    uri = URI.parse(configuration.verify_url + "/#{project_id}/assessments" + '?' + query)
+    http_instance = http.new(uri.host, uri.port)
+    http_instance.read_timeout = http_instance.open_timeout = timeout
+    http_instance.use_ssl = true if uri.port == 443
+    request = Net::HTTP::Post.new(uri.request_uri)
+    request['Content-Type'] = 'application/json; charset=utf-8'
+    request.body = JSON.generate(body)
+    JSON.parse(http_instance.request(request).body)
+  end
```
Comment (grosser) [inline lib/recaptcha.rb:175]: this seems very copy-paste, can  we extract a method to reuse ?

File: lib/recaptcha/configuration.rb [hunk 5]
```diff
@@ -45,6 +48,11 @@ def initialize #:nodoc:
 
       @secret_key = ENV['RECAPTCHA_SECRET_KEY']
       @site_key = ENV['RECAPTCHA_SITE_KEY']
+
+      @enterprise = ENV['RECAPTCHA_ENTERPRISE'] == 'true'
+      @api_key = ENV['RECAPTCHA_ENTERPRISE_API_KEY']
```
Comment (grosser) [inline lib/recaptcha/configuration.rb:53]: should this be enterprise_api_key ?

File: lib/recaptcha/configuration.rb [hunk 6]
```diff
@@ -45,6 +48,11 @@ def initialize #:nodoc:
...
+      @project_id = ENV['RECAPTCHA_ENTERPRISE_PROJECT_ID']
```
Comment (grosser) [inline lib/recaptcha/configuration.rb:54]: ... and enterprise_project_id ?

File: lib/recaptcha.rb (hunk 4 above)
Comment (jcmfernandes) [reply lib/recaptcha.rb:172]: I agree. Refactoring.

File: lib/recaptcha/configuration.rb (hunk 5 above)
Comment (jcmfernandes) [reply lib/recaptcha/configuration.rb:53]: Sure!

File: README.md [hunk 7]
```diff
@@ -68,6 +68,14 @@ export RECAPTCHA_SITE_KEY   = '6Lc6BAAAAAAAAChqRbQZcn_yyyyyyyyyyyyyyyyy'
 export RECAPTCHA_SECRET_KEY = '6Lc6BAAAAAAAAKN3DRm6VA_xxxxxxxxxxxxxxxxx'
 ```
 
+If you have an Enterprise API key:
+
+```shell
+export RECAPTCHA_ENTERPRISE            = 'true'
```
[the remark this answers is not in the document]
Comment (jcmfernandes) [reply README.md:74]: I personally prefer the current approach as it allows to decouple enabling this functionality from secret-setting, something that at hopin is performed by our devops team.

One option could be
```rb
def enterprise
  @enterprise || (secret_key.nil? && !enterprise_api_key.nil? && !enterprise_project_id.nil?)
end
```

what do you think?

**Comment by susarla12345** [comment]
@jcmfernandes I have started using this branch for enterprise recaptcha, But i am getting this error from the function verify_via_api_call_enterprise and the error is

{"error"=>{"code"=>403, "message"=>"The provided cloud project number does not own the given recaptcha key", "status"=>"PERMISSION_DENIED"}}

I think it is because of the enterprise api key i am using. I assume api key is the one that we get from api & services -> credentials

can you please help me with this

**Comment by jcmfernandes** [comment]
> @jcmfernandes I have started using this branch for enterprise recaptcha, But i am getting this error from the function verify_via_api_call_enterprise and the error is
> 
> {"error"=>{"code"=>403, "message"=>"The provided cloud project number does not own the given recaptcha key", "status"=>"PERMISSION_DENIED"}}
> 
> I think it is because of the enterprise api key i am using. I assume api key is the one that we get from api & services -> credentials
> 
> can you please help me with this

That's is correct, you need to create an API key in "APIs & Services -> Credentials -> API Keys". To start, make sure you don't set any restrictions. The project ID is in the following form:

image

**Comment by grosser** [comment]
thx! 5.8.0

**Comment by jcmfernandes** [comment]
Thank you @grosser! 🙇

**Comment by jcmfernandes** [comment]
Hey @pomartel getting back to you as we are seeing the same rise in requests being detected as bots. Our score distribution shifted quite dramatically towards the lower end of the scale. Are you still seeing the same issue?

**Comment by pomartel** [comment]
@jcmfernandes Yes, I also saw an increase of false positive at first and I had to scale back the minimum passing score to 0.2. However within a couple of weeks, the score distribution went up and now it seems pretty good overall. I think that the machine learning needs time to adjust and learn the different patterns of your website.

**Comment by jcmfernandes** [comment]
Thanks for the feedback @pomartel! :raised_hands:

**Comment by susarla12345** [comment]
@jcmfernandes, i get this as response when enterprise api call is done. Am i doing something wrong. Can you please help
`"tokenProperties"=>{"valid"=>false, "invalidReason"=>"INVALID_REASON_UNSPECIFIED", "hostname"=>"", "action"=>""}`

**Comment by jcmfernandes** [comment]
Hard to say @susarla12345. Are you using release 5.8.0?

**Pull request merged**